bindings/rust/bench/benches/memory.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use criterion::{
+    criterion_group, criterion_main,
+    measurement::{Measurement, ValueFormatter},
+    BenchmarkGroup, Criterion,
+};
+use rand::thread_rng;
+use rand_distr::{Distribution, Normal};
+use std::{fs::read_to_string, path::Path, time::Duration};
+
+fn get_bytes_from_snapshot(name: &str, i: i32) -> i32 {
+    // number of bytes in snapshot starts on 8th line, 12th character
+    read_to_string(format!("target/memory/{name}/{i}.snapshot"))
+        .unwrap()
+        .lines()
+        .nth(7)
+        .unwrap()[11..]
+        .parse()
+        .unwrap()
+}
+
+/// Get the difference in bytes between two snapshots, which is memory of the
+/// `i`th TlsBenchHarness (client and server)
+fn get_bytes_diff(name: &str, i: i32) -> i32 {
+    get_bytes_from_snapshot(name, i + 1) - get_bytes_from_snapshot(name, i)
+}
+
+/// Read in memory snapshots and give Criterion memory values to analyze and plot
+fn read_library(group: &mut BenchmarkGroup<'_, Bytes>, name: &str) {
+    if !Path::new("target/memory/").is_dir() {
+        panic!("need to run memory-bench.sh first");
+    }
+
+    let mut i = 0;
+    let normal = Normal::new(0.0, 10000.0).unwrap();
+    group.bench_function(name, |b| {
+        // iter_custom requires closure returning time taken for `iters` iterations
+        // read in `iters` memory snapshots and return sum of memory taken
+        b.iter_custom(|iters| {
+            let mut sum = 0;
+            for _ in 0..iters {
+                sum += get_bytes_diff(name, i) + normal.sample(&mut thread_rng()) as i32; // add jitter for plotting
```

Comment:
With the way the Criterion plots are made as violin plots, with no jitter the distribution of memory is too narrow and plots are infinitely narrow/invisible. We need to add jitter to see them on the plot. I know this isn't ideal and a different type of graph would work better, but since we want everything in one report in Criterion, we have to adapt the memory data to Criterion, and I think it's worth the tradeoff to have all the data in one place in the HTML report.

Reply:
The problem is that the plotting is narrow enough to be 0 pixels wide, and it also confuses the Criterion statistical analysis because for one of the libraries without jitter the memory usage is exactly the same for all 100 handshakes which causes a divide by 0. Right now, in the graph, it's already pretty narrow, but we could always make it even more narrow by decreasing the jitter:
![image](https://github.com/aws/s2n-tls/assets/76919968/d9a9fbf7-ada6-47cd-b609-e2146b09d0d4)
When showing this graph anywhere, we should definitely call out that we add jitter, but again, I think the jitter's worth the time saved from not making a graph and having it in the Criterion ecosystem.